internal/app/machined/pkg/controllers/cri/image_cache_config.go

```diff
@@ -326,3 +440,105 @@ func (ctrl *ImageCacheConfigController) getImageCacheRoot(ctx context.Context, r
 
 	return optional.Some(targetPath), true, nil
 }
+
+func (ctrl *ImageCacheConfigController) copyImageCache(ctx context.Context, logger *zap.Logger, source, target string) error {
+	logger.Info("copying image cache", zap.String("source", source), zap.String("target", target))
+
+	if ctrl.DisableCacheCopy {
+		// used for testing
+		return nil
+	}
+
+	var bytesCopied int64
+
+	if err := filepath.WalkDir(source, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("error walking source directory: %w", err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		relPath, err := filepath.Rel(source, path)
+		if err != nil {
+			return fmt.Errorf("error getting relative path: %w", err)
+		}
+
+		targetPath := filepath.Join(target, relPath)
+
+		info, err := d.Info()
+		if err != nil {
+			return fmt.Errorf("error getting file info: %w", err)
+		}
+
+		// we only support directories and files
+		switch {
+		case info.Mode().IsDir():
+			if err := os.MkdirAll(targetPath, 0o755); err != nil {
+				return fmt.Errorf("error creating directory: %w", err)
+			}
+
+			return nil
+		case info.Mode().IsRegular():
+			bytesCopied += info.Size()
+
+			return copyFileSafe(path, targetPath)
+		default:
+			return fmt.Errorf("unsupported file type %s: %s", info.Mode(), path)
+		}
+	}); err != nil {
+		return fmt.Errorf("error copying image cache: %w", err)
+	}
+
+	logger.Info("image cache copied", zap.String("size", humanize.IBytes(uint64(bytesCopied))))
+
+	ctrl.cacheCopyDone = true
+
+	return nil
+}
+
+func copyFileSafe(src, dst string) error {
+	srcStat, err := os.Stat(src)
+	if err != nil {
+		return fmt.Errorf("error getting source file info: %w", err)
+	}
+
+	dstStat, err := os.Stat(dst)
+	if err == nil && srcStat.Size() == dstStat.Size() {
```

Comment:
two files could haave the same size, but hashing is expensive

Reply:
that would be outside of our control I guess, we just try to skip copying what is already there